### PR TITLE
New buffer closing scheme

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/lantern-internal#3336

Find and fixes another theoretical deadlock on closing stream.

### CODE REVIEW

File: receive_buffer.go [hunk 1]
```diff
@@ -39,20 +37,19 @@ func newReceiveBuffer(defaultHeader []byte, ack io.Writer, pool BufferPool, wind
...
-	closed := buf.closed
-	if closed {
-		buf.mx.RUnlock()
+	select {
+	case <-buf.closed:
 		return
+	case buf.in <- frame:
+		// okay
 	}
-	buf.in <- frame
```
Comment (oxtoacart) [inline receive_buffer.go]: This was the bug. If for some reason we weren't reading from `buf.in`, this could hang and block `buf.mx`, which in turn would block `buf.close`. Reading from `buf.in` can block if `buf.in` is full and we're blocked on sending an ACK. That's not a high percentage thing, which would explain why this doesn't manifest more severely and more quickly, but over time we might be seeing these building up.

The solution is to use a channel rather than a Writer for the acks. Under the covers, that's what we were using anyway, so this just makes the interface a bit more precise and gives us the opportunity to interrupt on close by using a `select`.

File: receive_buffer.go [hunk 2]
```diff
@@ -132,24 +129,18 @@ func (buf *receiveBuffer) read(b []byte, deadline time.Time) (totalN int, err er
...
 
-func (buf *receiveBuffer) sendACK() {
-	buf.mx.RLock()
-	closed := buf.closed
-	buf.mx.RUnlock()
-	if closed {
-		// Don't bother acking
+func (buf *receiveBuffer) doSendACK(unacked int) {
+	select {
+	case <-buf.closed:
```
Comment (oxtoacart) [inline receive_buffer.go]: Don't block on sending ACK if we've closed

File: receive_buffer.go [hunk 3]
```diff
@@ -76,16 +73,15 @@ func (buf *receiveBuffer) read(b []byte, deadline time.Time) (totalN int, err er
...
 			// Read next frame, continue loop
-			if !open && frame == nil {
-				// we've hit the end
-				err = io.EOF
-				buf.ackIfNecessary()
-				return
-			}
 			buf.onFrame(frame)
 			continue
+		case <-buf.closed:
```
Comment (oxtoacart) [inline receive_buffer.go]: We now check buf.closed instead of buf.in, since we can't close the latter anymore.

File: receive_buffer.go [hunk 4]
```diff
@@ -106,24 +102,29 @@ func (buf *receiveBuffer) read(b []byte, deadline time.Time) (totalN int, err er
 				return
 			}
 			timer := time.NewTimer(deadline.Sub(now))
+			stopTimer := func() {
+				if !timer.Stop() {
```
Comment (oxtoacart) [inline receive_buffer.go]: Drive-by, noticed that I wasn't closing the timer correctly.

n.b. I hate this API, `timer.Stop()` should just do the right thing.

File: sendbuffer.go [hunk 5]
```diff
@@ -34,7 +33,7 @@ type sendBuffer struct {
 	writeTimer     *time.Timer
 }
 
-func newSendBuffer(defaultHeader []byte, w io.Writer, windowSize int) *sendBuffer {
+func newSendBuffer(defaultHeader []byte, out chan []byte, windowSize int) *sendBuffer {
```
Comment (oxtoacart) [inline sendbuffer.go]: sendbuffer didn't have any closing problems, but I figured I'd keep the API consistent by passing in the out channel here rather than an io.Writer.

File: sendbuffer.go [hunk 6]
```diff
@@ -43,24 +42,30 @@ func newSendBuffer(defaultHeader []byte, w io.Writer, windowSize int) *sendBuffe
...
+
 	sendRST := false
 	defer func() {
 		if sendRST {
 			// Send an RST frame with the streamID
-			w.Write(withFrameType(buf.defaultHeader, frameTypeRST))
+			select {
+			case out <- withFrameType(buf.defaultHeader, frameTypeRST):
+				// okay
+			case <-closeTimedOut:
```
Comment (oxtoacart) [inline sendbuffer.go]: I don't think we had a problem here, but it's better to be safe than sorry and not block on writing if we've closed.

File: sendbuffer.go [hunk 7]
```diff
@@ -152,6 +157,8 @@ func (buf *sendBuffer) close(sendRST bool) {
 	default:
 		// close already requested, ignore
 	}
-	buf.writeTimer.Stop()
+	if !buf.writeTimer.Stop() {
```
Comment (oxtoacart) [inline sendbuffer.go]: Drive-by, closing timer correctly.

File: go.mod [hunk 8]
```diff
@@ -11,17 +11,18 @@ require (
 	github.com/getlantern/fdcount v0.0.0-20190912142506-f89afd7367c4
 	github.com/getlantern/filepersist v0.0.0-20160317154340-c5f0cd24e799 // indirect
 	github.com/getlantern/golog v0.0.0-20190830074920-4ef2e798c2d7
-	github.com/getlantern/idletiming v0.0.0-20200206172754-f356a3557cf3
+	github.com/getlantern/idletiming v0.0.0-20200228204104-10036786eac5
```
Comment (oxtoacart) [inline go.mod]: Drive-by, updated to latest idletiming, dramatically reduces dependencies.

File: lampshade_test.go [hunk 9]
```diff
@@ -473,6 +473,8 @@ func TestOldClientsWithoutTimestamp(t *testing.T) {
 }
 
 func testConnBasicFlow(t *testing.T) {
+	closeTimeout = 1 * time.Second
```
Comment (oxtoacart) [inline lampshade_test.go]: Makes test complete faster

File: receive_buffer.go [hunk 10]
```diff
@@ -78,8 +86,8 @@ func (buf *receiveBuffer) read(b []byte, deadline time.Time) (totalN int, err er
 		select {
 		case frame, open := <-buf.in:
 			// Read next frame, continue loop
-			if !open && frame == nil {
```
Comment (oxtoacart) [inline receive_buffer.go]: Checking for frame == nil is unnecessary, if the channel is closed we won't have read anything from it.

File: receive_buffer.go [hunk 11]
```diff
@@ -106,18 +114,23 @@ func (buf *receiveBuffer) read(b []byte, deadline time.Time) (totalN int, err er
...
+				stopTimer()
 				buf.ackIfNecessary()
 				return
 			case frame, open := <-buf.in:
 				// Read next frame, continue loop
-				timer.Stop()
-				if !open && frame == nil {
+				if !open {
 					// we've hit the end
+					stopTimer()
```
Comment (oxtoacart) [inline receive_buffer.go]: Drive-by, stop the timer

File: receive_buffer.go (hunk 4 above)
Comment (hwh33) [reply receive_buffer.go]: Agreed, `timer.Stop` should drain the channel.  I can't think of a good reason that it wouldn't.

However, I don't think draining the channel is necessary here.  `timer` is getting thrown away anyway, right?  So it doesn't matter if there's still a value on its channel.

File: receive_buffer.go (hunk 1 above)
Comment (hwh33) [reply receive_buffer.go]: I think this can be simplified.  [Here's a PR to demonstrate](https://github.com/getlantern/lampshade/pull/50/files).  As far as I can tell, it is logically equivalent, though I could be missing something.

Comment (oxtoacart) [reply receive_buffer.go]: Hehe, I actually tried that along the way, but the problem is that it can cause use to lose some of the buffered data. Consider this sequence:

1. Buffer
2. Read from buffer
3. Buffer
4. Close
5. Attempt to read from buffer

On step 5, we want to get what was buffered in step 3 but since we've already closed, it's possible that we won't (`select` is non-deterministic if multiple cases are satisfiable).

So, I had to go back to using a lock, and the only way I could make it work was by adding a timeout.

File: receive_buffer.go (hunk 4 above)
Comment (oxtoacart) [reply receive_buffer.go]: Ahh, you made me realize that I don't even need to call stop(), because there's no goroutine to stop. I'll fix.

Comment (oxtoacart) [reply receive_buffer.go]: Fixed.

**Comment by oxtoacart** [comment]
Hmm, I tried this out on `fp-donyc3u16-20200226-042` and we still seem to have some streams stuck in closing. Will dig into it....

**Comment by oxtoacart** [comment]
Okay, I added some instrumentation to see where exactly we're getting hung up, and it's definitely in closing the send buffers:

```
Mar  3 01:02:52 fp-donyc3u16-20200226-042 http-proxy[23716]: DEBUG lampshade: session.go:47 Streams     Open: 232   Closing: 94   Closed: 2994
Mar  3 01:02:52 fp-donyc3u16-20200226-042 http-proxy[23716]: DEBUG lampshade: session.go:48 Receive Buffers        Closing: 0
Mar  3 01:02:52 fp-donyc3u16-20200226-042 http-proxy[23716]: DEBUG lampshade: session.go:49 Send Buffers           Closing: 94
```

File: sendbuffer.go [hunk 12]
```diff
@@ -1,18 +1,26 @@
...
 var (
-	closeTimeout = 30 * time.Second
+	closeTimeout = uint64(30 * time.Second)
 )
 
+func getCloseTimeout() time.Duration {
+	return time.Duration(atomic.LoadUint64(&closeTimeout))
+}
+
+func setCloseTimeout(newTimeout time.Duration) {
```
Comment (oxtoacart) [inline sendbuffer.go]: Added this so that I can change the timeout during testing to speed things up

File: sendbuffer.go [hunk 13]
```diff
@@ -112,36 +131,47 @@ func (buf *sendBuffer) sendLoop(w io.Writer) {
 }
 
 func (buf *sendBuffer) send(b []byte, writeDeadline time.Time) (int, error) {
-	buf.muClosing.RLock()
-	n, err := buf.doSend(b, writeDeadline)
-	buf.muClosing.RUnlock()
-	return n, err
+	for {
```
Comment (oxtoacart) [inline sendbuffer.go]: Need to do this trick here so that we don't deadlock on trying to write to `buf.in` while also trying to close it.

File: sendbuffer.go [hunk 14]
```diff
@@ -112,36 +131,47 @@ func (buf *sendBuffer) sendLoop(w io.Writer) {
...
+		return true, 0, ErrTimeout
 	}
-	buf.writeTimer.Reset(writeDeadline.Sub(now))
 	select {
 	case buf.in <- b:
-		return len(b), nil
-	case <-buf.writeTimer.C:
-		return 0, ErrTimeout
+		return true, len(b), nil
+	case <-time.After(writeDeadline.Sub(now)):
```
Comment (oxtoacart) [inline sendbuffer.go]: I'm just using time.After rather than a permanent writeTimer here.

**Comment by oxtoacart** [comment]
Okay, with the latest changes, I'm no longer seeing the leak of streams in "closing" state.

```
Mar  3 02:54:22 fp-donyc3u16-20200226-042 http-proxy[24037]: DEBUG lampshade: session.go:46 Sessions    Open: 23   Closing: 0   Closed: 35   Recv Loops: 23   Send Loops: 23
Mar  3 02:54:22 fp-donyc3u16-20200226-042 http-proxy[24037]: DEBUG lampshade: session.go:47 Streams     Open: 118   Closing: 0   Closed: 861
Mar  3 02:54:22 fp-donyc3u16-20200226-042 http-proxy[24037]: DEBUG lampshade: session.go:48 Receive Buffers        Closing: 0
Mar  3 02:54:22 fp-donyc3u16-20200226-042 http-proxy[24037]: DEBUG lampshade: session.go:49 Send Buffers           Closing: 0
```

**Comment by forkner** [comment]
this seems okay to me by eye.

**Comment by oxtoacart** [comment]
Thanks @forkner!
